railgun/internal/manager/controllerdef.go

```diff
@@ -199,49 +199,60 @@ func setupControllers(logger logr.Logger, mgr manager.Manager, proxy proxy.Proxy
 		Disabling Ingress controller`)
 	}
 
-	kongClusterPluginGVR := schema.GroupVersionResource{
-		Group:    konghqcomv1.SchemeGroupVersion.Group,
-		Version:  konghqcomv1.SchemeGroupVersion.Version,
-		Resource: "kongclusterplugins",
-	}
-	if ctrlutils.CRDExists(mgr.GetClient(), kongClusterPluginGVR) {
-		logger.Info("kongclusterplugins.configuration.konghq.com v1beta1 CRD available on cluster.")
-		controller := ControllerDef{
-			IsEnabled: &c.KongClusterPluginEnabled,
-			Controller: &configuration.KongV1KongClusterPluginReconciler{
-				Client:           mgr.GetClient(),
-				Log:              ctrl.Log.WithName("controllers").WithName("KongClusterPlugin"),
-				Scheme:           mgr.GetScheme(),
-				Proxy:            proxy,
-				IngressClassName: c.IngressClassName,
-			},
+	if c.KongClusterPluginEnabled == util.EnablementStatusEnabled {
+		kongClusterPluginGVR := schema.GroupVersionResource{
+			Group:    konghqcomv1.SchemeGroupVersion.Group,
+			Version:  konghqcomv1.SchemeGroupVersion.Version,
+			Resource: "kongclusterplugins",
+		}
+
+		if ctrlutils.CRDExists(mgr.GetClient(), kongClusterPluginGVR) {
+			logger.Info("kongclusterplugins.configuration.konghq.com v1beta1 CRD available on cluster.")
+			controller := ControllerDef{
+				IsEnabled: &c.KongClusterPluginEnabled,
+				Controller: &configuration.KongV1KongClusterPluginReconciler{
+					Client:           mgr.GetClient(),
+					Log:              ctrl.Log.WithName("controllers").WithName("KongClusterPlugin"),
+					Scheme:           mgr.GetScheme(),
+					Proxy:            proxy,
+					IngressClassName: c.IngressClassName,
+				},
+			}
+			controllers = append(controllers, controller)
+		} else {
+			logger.Info(`kongclusterplugins.configuration.konghq.com v1beta1 CRD not available on cluster.
+		Disabling KongClusterPlugin controller`)
```

Comment:
So if the user specifically had this controller enabled, not having the CRD is technically a failure right?

Should we instead do this?

```suggestion
			message := fmt.Sprintf("%s CRD not available on the cluster", kongClusterPluginGVR.String())
			if c.KongClusterPluginEnabled == util.EnablementStatusEnabled {
				return nil, fmt.Errorf(message)
			}
			logger.Infof("%s: disabling controller", message)
```

This would require making changes to the `setupControllers` signature to include an error but since we're already in here trying to improve the CRD and controller setup logic perhaps this is as good a time as any?

Reply:
setupControllers does not have that return syntax yet.